README.md

[![Build status — Travis-CI][travis-icon]][travis]
[![changelog][release-icon]][changelog]
[![js-semistandard-style][semi-icon]][semi]
[![Accessibility testing][pa11y-icon]][pa11y-ci]

# openEssayist-slim

Web frontend to the essay analyser and summariser — [pyEssayAnalyser][py].

Part of the [SafeSEA][] project.
(Grants: _[Open University][ou-grant], [Oxford University][ox-grant]_.)
(_Research [papers on ORO][oro-ss]. [YouTube demo][yt-demo]_)

Built on [Slim PHP framework][slim].

* GitHub: [SAFeSEA/openEssayist-slim][gh]
* GitHub: [IET-OU/openEssayist-slim][gh-iet]
* GitHub: [SAFeSEA/pyEssayAnalyser][gh-py]

## Authentication

Newer OpenEssayist servers can optionally use [OU-SAMS][] authentication,
with a pre-defined list of [admin OUCUs][admin].

## Release notes

* See: [CHANGELOG][]

## Redhat preparation

```sh
composer redhat-check
composer redhat-install   # yum install php php-fpm php-mcrypt php-zip httpd mod_fcgid
# cp -n app/_data/php.fastcgi /var/www/cgi-bin/
#   ...
```

## Install .. Test

```sh
composer prepare    # Edit: 'app/config.php'
                    # OU only - edit: '.env'
composer install
composer npm-install
composer copy-nginx  # Only on Redhat 7.
composer chown       # Only on Redhat 7.

composer build
composer test
```

## Database

To create and seed the [MySQL database tables][db], start typing:

```sh
composer cli-help
app/cli.php --create-tables
```

## Auto-assign student to group

Edit the [groups][] PHP `object` (`array`) in the `app/config.php` configuration file, then run `app/cli.php`:

```sh
sudo vi app/config.php +78
sudo app/cli.php --seed-groups
```

The student then uses a link of this form, to add themselves to the group on their first visit:

```
https://my-openessayist.open.ac.uk/samslogin?group={NAME}
```

---

> To check:

* https://github.com/silentworks/sharemyideas
* https://github.com/Tieno/SlimPackage
* https://github.com/codeguy/Slim-Extras
* [Using phpactiverecord with slim..](http://silentworks.co.uk/blog/development/using-phpactiverecord-with-slim-framework.html)

---

* Original developer: [vanch3d][].

## GDPR

Details of GDPR / privacy fixes can be found in [Bug #1][] (Was [Bug #38][]).

---
[©][c] 2013-2019 [The Open University][ou]. ([Institute of Educational Technology][iet])

[c]: https://www.open.ac.uk/copyright "Copyright © 2013-2019 The Open University (IET). All rights reserved."
[ou]: https://www.open.ac.uk/
[iet]: https://iet.open.ac.uk/

[vanch3d]: https://github.com/vanch3d "Original developer: Nicolas Van Labeke (vanch3d)"
[changelog]: https://github.com/IET-OU/openEssayist-slim/blob/3.x/CHANGELOG.md

[py]: https://github.com/SAFeSEA/pyEssayAnalyser
[gh]: https://github.com/SAFeSEA/openEssayist-slim "Original"
[gh-iet]: https://github.com/IET-OU/openEssayist-slim "Fork"
[gh-py]: https://github.com/SAFeSEA/pyEssayAnalyser "Python"
[travis]: https://travis-ci.org/IET-OU/openEssayist-slim "IET-OU / openEsasyist-slim"
[travis-icon]: https://travis-ci.org/IET-OU/openEssayist-slim.svg
[travis-ss]:  https://travis-ci.org/SAFeSEA/openEssayist-slim "SafeSEA / openEssayist-slim"
[travis-ss-icon]: https://api.travis-ci.org/SAFeSEA/openEssayist-slim.svg
    "Build status – Travis-CI (PHP)"
[semi]: https://github.com/Flet/semistandard
[semi-icon]: https://img.shields.io/badge/code_style-semistandard-brightgreen.svg?_style=flat-square
    "Javascript coding style — 'semistandard'"
[pa11y-ci]: https://github.com/pa11y/pa11y-ci
    "Automated accessibility testing - via 'pa11y-ci' (work-in-progress)"
[pa11y-icon]: https://img.shields.io/badge/accessibility-pa11y--ci-blue.svg
[release-icon]: https://img.shields.io/github/release/IET-OU/openEssayist-slim.svg "Changelog: latest GitHub release"
[slim]: https://docs.slimframework.com/ "Slim PHP Framework v2"
[db]: https://github.com/IET-OU/openEssayist-slim/blob/3.x/app/_data/openessayist-schema.sql#L24 "SQL database schema"
[ou-sams]: https://github.com/IET-OU/openEssayist-slim/blob/3.x/composer.json#L45 "Composer: use the 'iet-ou/sams-c-auth' library"
[admin]: https://github.com/IET-OU/openEssayist-slim/blob/3.x/config/app.DIST.php#L62 "Configuration: 'admin_oucu_list' => [ .. ]"
[groups]: https://github.com/IET-OU/openEssayist-slim/blob/3.x/config/app.DIST.php#L82 "Configuration: 'groups' => [ .. ]"

[safesea]: https://www.open.ac.uk/researchprojects/safesea/
  "Supportive Automated Feedback for Short Essay Answers (SAFeSEA)."
[yt-demo]: https://youtu.be/7a3ATQPjpiM# "openEssayist Software Tool Demonstration, @ietou on YouTube"
[yt-intro]: https://youtu.be/a9l0ts1tgK4# "Introduction to openEssayist - Professor Denise Whitelock, @ietou on YouTube"
[oro-ss]: https://oro.open.ac.uk/cgi/search/archive/advanced?project_details_project_name=SafeSEA "'SafeSEA' on ORO (10 results)"
[oro-oe]: https://oro.open.ac.uk/cgi/search/archive/simple?meta=OpenEssayist& "'OpenEssayist' on ORO (8 results)"
[ou-grant]: http://gow.epsrc.ac.uk/NGBOViewGrant.aspx?GrantRef=EP/J005959/1
  "Supportive Automated Feedback for Short Essay Answers (SAFeSEA) (Open University, 2012-2014) [EPSRC grant: EP/J005959/1]"
[ox-grant]: http://gow.epsrc.ac.uk/NGBOViewGrant.aspx?GrantRef=EP/J005231/1
  "Supportive Automated Feedback for Short Essay Answers (SAFeSEA) (Oxford University, 2012-2014) [EPSRC grant: EP/J005231/1]"

[Bug #38]: https://github.com/SAFeSEA/openEssayist-slim/issues/38 "GDPR/privacy (SafeSEA)"
[bug #1]: https://github.com/IET-OU/openEssayist-slim/issues/1 "GDPR/privacy"

[End]: //.


CHANGELOG.md

[![Build status — Travis-CI][travis-icon]][travis]

# OpenEssayist changelog

← [README][]

## [Latest][]

 * _Date:  ~ 26-Oct-2018 ~ 12-Feb-2019 ~ to be confirmed!_
 * GDPR — privacy policy — `templates/pages/privacy.twig` ([Bug #1][] / Bug #38)
 * GDPR — only store student OUCU in database, not full-name ([Bug #1][])
 * GDPR — update `IET-OU/sams-c-auth` PHP library, and only allow registered students (and staff) to access, _not visitors / "prospects"_ ([Bug #1][])
 * Fix — remove non-brand-compliant OpenEssayist logo;
 * Add HTTP Basic Auth to apache/nginx configurations (_approval server_) — `app/_data/*.conf`;
 * Move route definitions to separate file — `app/utils/routes.php`;
 * Move `app/config.php` to `config/app.php` — Laravel convention;
 * Add '_Auto-assign student to group_' and '_GDPR_' section(s) to [README][];

## [Version 3.0.2][v3.0.2]

 * _Date:  ~ 12 June 2018_
 * Fix to 'control' the `task.code` field — make it readonly in the Twig/HTML template — _admin editor_ (Bug #37)
 * Fix, make the `task.wordcount` field `required` — _admin editor_;
 * Fix/set `isgroup` flag in `Login..::createUserSams()` (Bug #35)
 * Add [twig-lint][] to the Travis-CI test;
 * Add _Authentication_ section to [README][];

## [Version 3.0.1][v3.0.1-beta]

 * _Date:  ~ 19 February 2018_,
 * Add database column `users.visit_count`; record user's login (Bug #35),
 * Set `maxlength` to _30,000_ characters in [`app/config.php`][cfg],
 * Fix for user-roles in SAMS authentication,
 * Fix user-interface texts in templates ('_My assignments_', '_Feedback_'),
 * Fix handling of database connection errors,
 * Add automated accessibility testing — [pa11y-ci][] (Bug #36).

## [Version 3.0.0][v3.0.0-beta]

 * Date:  7 February 2018 _(Development: 10 Jan - 6 Feb.)_
 * Tag:   [3.0.0-beta][v3.0-co]  (Describe: `v2.6-111-g71a6a5f`)
 * Notes: [Google Docs][]
 * Built on:  PHP 7.0 / Slim framework v 2.2.0 / MySQL

Below are the __headline__ items. More details are in [Google Docs][].

### User-interface

 * Added configurable `maxlength` character limit to the draft-essay
 submission form — prevent analyser errors; reduce server-load (Bug #30)
 * Made error messages to user more user-friendly — for example, prefixed them with "_Sorry. ..._",
 * _(No other new user-interface features/functions — apart from SAMS authentication below.)_

### PHP

 * Added _OU-SAMS_-based authentication and account creation — via separate library (Bug #27)
 * Fix PHP dependencies - upgrade from `dev-master` to released versions, in `composer.json` and `composer.lock` (Bug #20)
 * Configuration — rename `app/config.php` to [`app/config.DIST.php`][cfg],
 * Configuration — make `rd_save_path`, `analyser_url` etc. configurable (Bug #23)
 * Added [Travis-CI][] continuous integration, with PHP linting [QA] (maintainability) (Bug #20)
 * Moved core _EssayAnalyser_ wrapper code to separate class (maintainability) (Bug #32)
 * Moved a lot of `require` and `Twig` configuration from front controller
 (`index.php`) to separate PHP files (maintainability) (Bug #31)
 * Moved database connection code from `app/config.php` to separate class (Bug #31)
 * Broke model classes into separate files (find-ability),
 * Added missing logging; improved error-handling (debugging aid, analytics),
 * Added _status_ / _uptime_ controller — shows status of backend analyser service (as well as front),
 * Eased and documented installation and setup, via [README][], Composer `scripts`.

### Database

 * Database schema fixes and edits - add dates/times, reduce column lengths (Bug #25)
 * Add analysis start/end time columns (Bug #25)
 * Prevent Web application from automatically creating database, tables, seeding test/example data (Bug #24)
 * New `app/cli.php` to create tables, seed with minimal/ configurable user, group and task data,
 * Configuration/ connection fixes to ensure data is sent/stored as `UTF-8` in database (Bug #26).

### Javascript

 * Javascript moved from `base.html.twig` and `site.twig` templates to separate JS files (maintainability, performance),
 * Javascript moved from some `drafts/view.*.twig` templates to separate JS files,
 * Added `semistandard` linting to [Travis-CI][] for 'openessayist' Javascripts — lots of fixes [QA]
 * Third-party Javascript/CSS versions documented in [`package.json`][pkg],
 * Transferred more key 3rd-party Javascript libraries to CDN (performance),
 * Centralised versioning of 3rd-party Javascript — via `VersionedAsset` PHP class (Bug #33)
 * Upgraded jQuery from 1.9.1 to 2.2.4 _(can't upgrade to 3.x - Bootstrap bug?)_
 * Added Google Analytics.

### Miscellaneous

 * Add missing Open University copyright statements — in PHP, Javascript, user-interface, `composer.json` (Bug #20)
 * Added _Nicolas Van Labeke_ (`vanch3d`) to `composer.json`
 * Removed legacy `composer.phar` binary from Git; removed `./openessayist.sql.zip` (2.6 MB)
 * Moved unused Javascript from Twig templates to separate JS file(s),
 * Added test data — sourced from [Project Gutenberg][pg] and [ORO][],
 * Documentation, changelog (Bug #34).


---
## [Version 2.7](https://github.com/IET-OU/openEssayist-slim/releases/tag/v2.7)

 * Date:   24 February 2014.
 * Tag:    _v2.7_ ("_v2.7 beta_")
 * Commit: [IET-OU/openEssayist-slim `7a3d3861`][v2.7-co]

> version 2.7 - ready for Phase II Evaluation

## [Version 2.6](https://github.com/IET-OU/openEssayist-slim/releases/tag/v2.6)

 * Date:   27 September 2013.
 * Tag:    _v2.6_
 * Commit: [IET-OU/openEssayist-slim `e6f00deb`][v2.6-co]

> version 2.6 - ready for September evaluation

## [Version 2.4](https://github.com/IET-OU/openEssayist-slim/releases/tag/v2.4)

 * Date:   16 August 2013.

> version 2.4 - August User Testing

_[...]_

## [Version 2](https://github.com/IET-OU/openEssayist-slim/releases/tag/v2_user_testing)

 * Date:   15 May 2013.
 * Tag:    _v2_user_testing_

> Version 2 of openEssayist, ready for user testing

---
## Previous versions

* v2_user_testing, v2.2, v2.3 — 15 May to 30 July 2013.

app/config.DIST.php
<?php
/**
 * OpenEssayist - database and application configuration.
 *
 * $activeGroup contains key for current databases
 * $db contains all database sets available
 * Change $activeGroup value to one of the $db keys, to activate a different database.
 *
 * NOTES:
 *  - make sure that the DB defined in 'database' field is created in MySQL
 *  - make sure that the directory defined in 'logdir' is created (and contains the log files)
 */

$activeGroup = 'local';

// local database and logs
$db['local']['hostname'] = 'localhost';
$db['local']['username'] = '** EDIT ME **';
$db['local']['password'] = '** EDIT ME **';
$db['local']['database'] = 'openessayist';
$db['local']['dbProvider'] = 'mysql';
$db['local']['logdir'] = __DIR__ . '/../.logs';

// H810 archive database and logs
$db['H810']['hostname'] = 'localhost';
$db['H810']['username'] = '** EDIT ME **';
$db['H810']['password'] = '** EDIT ME **';
$db['H810']['database'] = 'openh810';
$db['H810']['dbProvider'] = 'mysql';
$db['H810']['logdir'] = __DIR__ . '/../log-h810';

// H817 archive database and logs
$db['H817']['hostname'] = 'localhost';
$db['H817']['username'] = '** EDIT ME **';
$db['H817']['password'] = '** EDIT ME **';
$db['H817']['database'] = 'openh817';
$db['H817']['dbProvider'] = 'mysql';
$db['H817']['logdir'] = __DIR__ . '/../log-h817';

// Other configuration.

$_oe_email_ = 'openessayist-techsupport@open.ac.uk'; // Was: 'nicolas.vanlabeke@open.ac.uk'

$openessayist_config = [
	'email' => $_oe_email_,
	'analyser_url' => 'http://127.0.0.1:8062',
	'rd_save_path' => __DIR__ . '/../_user_data/images/',
	'draft_maxlength_chars' => 30 * 1000, // Was: 4000, (set after Simon Cross briefing).
	'_max_word_count_' => 5000,
	'task_code_regex' => '^(TMA01|ICS)$',  // Regex for use in Javascript & PHP.
	'group_code_regex'=> '^(H810|VCS)$',

	// SAMS authentication - only enable for Open University.
	'sams_enable' => false,
	'sams_redirect' => true,
	'sams_password' => '** EDIT ME **',
	'sams_group_id' => 1,  // Set a default.
	// See :~ https://github.com/IET-OU/sams-c-auth/blob/master/src/SamsCAuth.php#L29
	// NOT:  /^(...|visitor)$/
	'sams_auth_regex' => '/^(staff|tutor|student)$/',
	// Accounts in this admin list have the "isadmin" and "isgroup" flags set on their account.
	'admin_oucu_list' => [ 'abc123', '** EDIT ME **' ],
	'analyser_version' => __DIR__ . '/../../openessayist-python/version.json',

	'google_analytics' => 'UA-3845152-23', // GA property within 'IET Sites'.
	'analytics_prefix' => '/local',        // Example: '/server-A', '/server-B'. ** EDIT ME **

	// Seed database ... 1+ users, 1+ groups, 1+ tasks.

	'users' => [
		'admin' => (object) [
			'username' => 'admin',
			'password' => '** EDIT ME **',
			'email'  => $_oe_email_,
			'name' => 'Admin 01',
			'isadmin' => true,
			'auth_type' => 'seed',
		],
	],

	'groups' => [
		'default' => (object) [
			'name' => 'Default',
			'code' => 'H810',  // Must be 'H810', or 'VSO'.
			'description' => 'A default group.',
		]
	],

	'tasks' => [
		'task-01' => (object) [
			'name' => 'Task 01',
			'code' => 'TMA01',  // Must be 'TMA01 or ICS'.
			'assignment' => 'A default assignment.', // Open by default.
		]
	]
];


// ------------------------------------------------------------------------
// The database connection code has been moved to `app/utils/DBConnection.php` (2-Feb-2018).
// End.
